text/0000-path-clarity.md

```diff
@@ -0,0 +1,731 @@
+[previous RFC]: https://github.com/rust-lang/rfcs/pull/2088
+
+## Migration experience
+
+The two stages are carefully designed to retain key properties of the module
+system and compilation throughout, and to minimize the introduction of hard
+errors across epochs. The only thing that you *must* do to opt in to the next
+epoch is change absolute paths to use the new format (starts with crate name,
+`crate`, `super`, or `self`).
+
+We can and should provide a high-fidelity `rustfix` tool that performs at least
+this change, and ideally lowers `pub` to `crate` when possible as well.
+
+In the error specified for invalid fully qualified paths, in which the compiler
+checks for a top-level module with the same name, the error should *also*
+suggest that the code may be coming from an older epoch, and point to (1) the
+`rustfix` tool to automate migration and (2) the ability to specify an older
+epoch in `Cargo.toml`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The most clear drawback here is that a new epoch is required. The RFC strives to
+make the transition as narrow as possible, both by introducing as few new errors
+as possible, and by providing additional mitigations through migration tooling
+and compiler diagnostics (as well as the epoch system itself, of course).
+
+Another drawback is that imports from within your crate become more verbose,
+since they require a leading `crate`. However, this downside is considerably
+mitigated if [nesting in `use`] is permitted.
+
+[nesting in `use`]: https://github.com/rust-lang/rfcs/issues/1400
+
+In previous discussions about deprecating `extern crate`, there were concerns
+about the impact on non-Cargo tooling, and in overall explicitness. This RFC
+fully addresses both concerns by leveraging the new, unambiguous nature of fully
+qualified paths.
+
+# Rationale and Alternatives
+[alternatives]: #alternatives
+
+The core rationale here should be clear given the detailed analysis in the
+motivation. The crucial insight of the design is that, by making absolute paths
+unambiguous about which crate they draw from, we can solve a number of
+confusions and papercuts with the module system.
+
+## Bike-sheddy choices
+
+There are a few aspects of this proposal that could be colored a bit differently
+without fundamental change.
+
+- Rather than `crate::top_level_module`, we could consider `extern::serde` or
+  something like it. That would come with some significant downsides, though.
+  - First, having paths typically start with a crate name, with `crate`
+    referring to the current crate, provides a *very simple* and easy to
+    understand model for paths---and its one that's pretty commonly used in other languages.
+  - Second, one benefit of `crate` is that it helps reduce confusion about paths
+    appearing in `use` versus references to names elsewhere. In particular, it
+    serves as a reminder that `use` paths are absolute.
+
```

Comment:
Another alternative might be to improve error messages.  

Using example from Motivation, rather than just saying "Use of undeclared type or module `futures`", as rustc does now, it could add a note to the effect of "...but I did find an extern crate of this name.  You can use an absolute path to refer to it as `::futures::Poll`, or you can bring it into the current scope with `use futures;`.

Reply:
The asymmetry between lib.rs and all other files doesn't just show up in the `extern crate;` declarations, it also appears when you do `mod foo; use foo::bar::Baz;`, or `use std::mem; use mem::swap;`. You can only do this in the crate root and as probably your first steps in Rust are by coding stuff inside the crate root and after that moving stuff from there into modules, it gives you this moment of surprise right at the start when you realize that submodules work differently.

Reply:
My point was that rather than revamping the module system, we could have simply provide better error messages, like we do in other circumstances when the compiler can reasonably guess what went wrong.   If we can't find module 'foo' in the current scope, but there is one in crate root, it's probably what the user wanted.

I don't think I've seen this option mentioned anywhere (at least lately). If it *had* been considered, maybe it should be mentioned among the alternatives along with why this isn't a good idea.